### PR TITLE
#61 Add infinity support to Dates and Numerics

### LINKED ISSUE #61: Add ability to set values for infinite numbers for dates and timestamps
## Background
Mainframes have various ways of representing dates. 

As the result we need to have assign special values to minimum and maximum dates, and also be able to specify these minimum and maximum dates.

## Feature
Add a new pattern, for mainframe dates.

## Example

```hocon
metadata {
  "pattern": "yyyyMMdd",
  "minus_infinity_symbol": "0",
  "minus_infinity_value": "10000101",
  "plus_infinity_symbol": "99999",
  "plus_infinity_value":  "99991231"
}
```

## Proposed Solution [Optional]
All new metadata keys are optional:
- `minus_infinity_symbol` - not used if not specified
- `minus_infinity_value` - a fixed default value is used
- `plus_infinity_symbol`- not used if not specified
- `plus_infinity_value`  - a fixed default value is used

Default values for minimum and maximum dates and timestamps can be used form here:
https://github.com/apache/spark/blob/070461cc673c3fc910e66d1cbf628632b558b48c/sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala#L323

### PR DESCRIPTION
**Pull request by Zejnilovic**
Closes #61 

Adds support or infinity replacements on Dates, Timestamps and Numeric fields. 

It works before the "spark built-in" infinity replacements that use `∞` and has no dependency on `allowInfinity` metadata field. 

New metadata fields are now supported are:
- minus_infinity_symbol
- minus_infinity_value
- plus_infinity_symbol
- plus_infinity_value

### CODE REVIEW

File: src/main/scala/za/co/absa/standardization/stages/TypeParser.scala [hunk 1]
```diff
@@ -681,4 +695,27 @@ object TypeParser {
   }
 }
 
+sealed trait InfinitySupport {
```
Comment (salamonpavel) [inline src/main/scala/za/co/absa/standardization/stages/TypeParser.scala:698]: why sealed trait?

File: project/build.properties [hunk 2]
```diff
@@ -1 +1 @@
-sbt.version=1.6.2
+sbt.version=1.10.2
```
Comment (benedeki) [inline project/build.properties:1]: Is this needed?

File: src/main/scala/za/co/absa/standardization/stages/TypeParser.scala (hunk 1 above)
Comment (benedeki) [reply src/main/scala/za/co/absa/standardization/stages/TypeParser.scala:698]: It could and perhaps should be in its own file even.

Comment (Zejnilovic) [reply src/main/scala/za/co/absa/standardization/stages/TypeParser.scala:698]: The only reason is that this is a library and there are not supposed to be used outside. It is supposed to be only used with our defined parsers.

File: project/build.properties (hunk 2 above)
Comment (Zejnilovic) [reply project/build.properties:1]: Maybe not. Is it an issue?

**Comment by Zejnilovic** [comment]
> Here is a problematic use case.
> 
> Config:
> 
> ```json
> {
>  "pattern": "yyyyMMdd", 
>  "minus_infinity_symbol": "0", 
>  "minus_infinity_value": "00010101", 
>  "plus_infinity_symbol": "99999999", 
>  "plus_infinity_value": "99991231"
> }
> ```
> 
> When `"minus_infinity_value": "10000101"` this works fine. But when `"minus_infinity_value": "00010101"` it fails to properly convery minus infinity.
> 
> I believe this is because this implementation tries to replace all '0' with value `00010101` _using the original type_, which converts '0' to `10101` since the original type is integer.
> 
> In order to have pure solution that covers this and similar use cases maybe we should do it like this (pseudocode):
> 
> ```
> IF col == plus_inf_symbol THEN plus_inf_value
> ELSE IF  col ==  minus_inf_symbol THEN minus_inf_value
> ELSE convert(col, pattern)
> ```
> 
> where `plus_inf_value` and `minus_inf_value` are precomputed dates.
> 
> It is more complicated, but covers edge cases like this.
> 
> ### Alternatively
> You can convert the input column to a string first, then do the replacement, and then do a conversion
> 
> ### In addition (optional, just thinking)
> Maybe we can consider specifying `minus_infinity_value` and `plus_infinity_value` always in some standard pattern, so that it can represent values that are outside of parsing pattern.
> 
> Say, this can solve weird cases like this:
> 
> ```json
> {
>  "pattern": "yyyy-MM", 
>  "minus_infinity_symbol": "0", 
>  "minus_infinity_value": "0001-01-15", 
>  "plus_infinity_symbol": "999999", 
>  "plus_infinity_value": "9999-12-31"
> }
> ```

My thoughts were that an integer or numeric columns in general would never come with zeros in the front (i.e. `01010001` or `00010101`) irrespective of the infinity support. Infinity support with a solution like that would solve for something that could never happen on simple input data and create an edge case. Thus did not want to include it. 

As per @yruslan  we will quickly confirm if this is need feature and think of a solution or if this is good enough.

File: src/main/scala/za/co/absa/standardization/stages/TypeParser.scala (hunk 1 above)
Comment (benedeki) [reply src/main/scala/za/co/absa/standardization/stages/TypeParser.scala:698]: I think the code of this trait is written in such a way, that outside usage doesn't really causes issue, if somebody finds it useful - unlikely 😉

Comment (benedeki) [reply src/main/scala/za/co/absa/standardization/stages/TypeParser.scala:698]: IMHO the "hiding" is important, when the class/trait have special handling needs or has access to more sensitive parts of the code.

Comment (Zejnilovic) [reply src/main/scala/za/co/absa/standardization/stages/TypeParser.scala]: Moved out

**Comment by yruslan** [comment]
I'm okay with the current implementation.

**Comment by kevinwallimann** [comment]
> > Here is a problematic use case.
> > Config:
> > ```json
> > {
> >  "pattern": "yyyyMMdd", 
> >  "minus_infinity_symbol": "0", 
> >  "minus_infinity_value": "00010101", 
> >  "plus_infinity_symbol": "99999999", 
> >  "plus_infinity_value": "99991231"
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > When `"minus_infinity_value": "10000101"` this works fine. But when `"minus_infinity_value": "00010101"` it fails to properly convery minus infinity.
> > I believe this is because this implementation tries to replace all '0' with value `00010101` _using the original type_, which converts '0' to `10101` since the original type is integer.
> > In order to have pure solution that covers this and similar use cases maybe we should do it like this (pseudocode):
> > ```
> > IF col == plus_inf_symbol THEN plus_inf_value
> > ELSE IF  col ==  minus_inf_symbol THEN minus_inf_value
> > ELSE convert(col, pattern)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > where `plus_inf_value` and `minus_inf_value` are precomputed dates.
> > It is more complicated, but covers edge cases like this.
> > ### Alternatively
> > You can convert the input column to a string first, then do the replacement, and then do a conversion
> > ### In addition (optional, just thinking)
> > Maybe we can consider specifying `minus_infinity_value` and `plus_infinity_value` always in some standard pattern, so that it can represent values that are outside of parsing pattern.
> > Say, this can solve weird cases like this:
> > ```json
> > {
> >  "pattern": "yyyy-MM", 
> >  "minus_infinity_symbol": "0", 
> >  "minus_infinity_value": "0001-01-15", 
> >  "plus_infinity_symbol": "999999", 
> >  "plus_infinity_value": "9999-12-31"
> > }
> > ```
> 
> My thoughts were that an integer or numeric columns in general would never come with zeros in the front (i.e. `01010001` or `00010101`) irrespective of the infinity support. Infinity support with a solution like that would solve for something that could never happen on simple input data and create an edge case. Thus did not want to include it.
> 
> As per @yruslan we will quickly confirm if this is need feature and think of a solution or if this is good enough.

Tbh, I think the argument doesn't make sense from a user's perspective. Why should a minus infinity symbol of 0 be standardized to a date 1000-01-01 and not 0001-01-01, or even 0000-01-01 for that matter. But we can tell users that the minimal supported date is 1000-01-01, so I'm also ok with the functionality.

**Review by kevinwallimann — APPROVED** [review]

**Review by yruslan — APPROVED** [review]

**Review by benedeki — APPROVED** [review]
LGTM

**Pull request merged**